File: scripts/lib/license.mjs
// Licence keys: mint, parse, verify.
//
// The app cannot phone home. scripts/audit-network.sh fails the build on
// URLSession, getaddrinfo and the TLS entry points, and docs/licensing.md sells
// that claim on the front page — so a key has to prove itself from nothing but
// the bytes it carries and a public key compiled into the binary. That rules out
// every activation server and leaves exactly one shape: sign a payload here,
// verify it there, never ask anyone.
//
// Ed25519 rather than RSA or an HMAC. An HMAC would put the minting key inside
// the app, where anyone could read it and issue their own. RSA is four times the
// bytes for no gain. What is left is a 64-byte signature, which is still long
// enough that a key is pasted or dropped and never typed — the UI is designed
// for that rather than surprised by it.
//
// None of this is a tamper defence and none of it should become one.
// apps/apple/LICENSE §1(c) already grants anyone the right to compile the app
// and run it with no key at all, so effort spent hardening this is effort spent
// against a door that is deliberately open. What it buys is a key that cannot be
// forged or altered — not a program that cannot be modified.
//
// Dependency-free (node builtins only), like the probes. apps/api implements the
// same format against WebCrypto, and the two must agree byte for byte — which is
// why the signature covers the ENCODED payload rather than the parsed object.
// JSON key order and whitespace then stop being anybody's problem.

import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  randomBytes,
  sign,
  verify,
} from "node:crypto";

/** Namespaces the format. Bumping it is how a v2 key stays distinguishable. */
export const KEY_PREFIX = "cup1";

/**
 * The public half of the signing key, in base64url.
 *
 * Lives here AND in apps/apple/Cupertino/License.swift, because the app cannot
 * import from this file and the Worker cannot import from the app. Two copies of
 * one constant is a drift risk with a nasty failure — every paid key refused at
 * once — so license.test.mjs reads the Swift file and asserts the two agree,
 * rather than trusting whoever edits one to remember the other.
 *
 * Not a secret — it only ever verifies. The private half lives in `.env` locally
 * and as a Worker secret in production, and is the one thing in this scheme that
 * must never leak and must never be lost: leaking it lets anyone issue keys,
 * losing it means no new keys can be issued for this major version ever again.
 */
export const PUBLIC_KEY = "_sGLrSm_Sg3bv2p0T8yfzelAvkEjAa2se9l2X4sgNA4";

/** Crockford base32, the ULID alphabet: no I, L, O or U, so it survives dictation. */
const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

const toBase64Url = (buffer) => Buffer.from(buffer).toString("base64url");
const fromBase64Url = (text) => Buffer.from(text, "base64url");

const privateKeyFrom = (privateKeyBase64) =>
  createPrivateKey({
    format: "der",
    key: Buffer.from(privateKeyBase64, "base64"),
    type: "pkcs8",
  });

const publicKeyFrom = (publicKeyBase64Url) =>
  createPublicKey({
    format: "jwk",
    key: { crv: "Ed25519", kty: "OKP", x: publicKeyBase64Url },
  });

/**
 * A ULID: 10 characters of millisecond timestamp, 16 of randomness.
 *
 * Sortable by issue time, which is what makes a revocation list readable rather
 * than a bag of opaque strings. 256 divides by 32, so a byte modulo the alphabet
 * is unbiased and needs no rejection loop.
 */
export const ulid = (now = Date.now()) => {
  let time = "";
  for (let remaining = now, i = 0; i < 10; i++) {
    time = CROCKFORD[remaining % 32] + time;
    remaining = Math.floor(remaining / 32);
  }
  let random = "";
  for (const byte of randomBytes(16)) random += CROCKFORD[byte % 32];
  return time + random;
};

/**
 * A fresh signing keypair.
 *
 * `privateKey` is PKCS#8 DER in base64 — one opaque string to paste into a
 * Worker secret, and the only thing that must never leak. `publicKey` is the raw
 * 32 bytes in base64url, because that is what CryptoKit's
 * `Curve25519.Signing.PublicKey(rawRepresentation:)` takes and what gets
 * compiled into the app.
 */
export const generateKeypair = () => {
  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  return {
    privateKey: privateKey.export({ format: "der", type: "pkcs8" }).toString("base64"),
    publicKey: publicKey.export({ format: "jwk" }).x,
  };
};

/** The public half of a private key, in the same form `generateKeypair` returns. */
export const publicKeyOf = (privateKeyBase64) =>
  createPublicKey(privateKeyFrom(privateKeyBase64)).export({ format: "jwk" }).x;

/**
 * Sign claims into a key string.
 *
 * `major` is the app major version the key unlocks: a key for 1 opens every 1.x
 * and nothing numbered 2, which is the whole of the per-major model in one
 * integer. `id` is the revocation handle and the only field a refund needs.
 */
export const mint = ({
  email,
  id = ulid(),
  issuedAt = new Date().toISOString(),
  major,
  privateKey,
}) => {
  const payload = toBase64Url(JSON.stringify({ id, email, major, issuedAt }));
  const signature = toBase64Url(sign(null, Buffer.from(payload), privateKeyFrom(privateKey)));
  return `${KEY_PREFIX}.${payload}.${signature}`;
};

/** Split a key into its parts without judging them. Throws on anything malformed. */
export const parse = (key) => {
  const parts = String(key ?? "")
    .trim()
    .split(".");
  if (parts.length !== 3) throw new Error("expected three dot-separated parts");
  const [prefix, payload, signature] = parts;
  if (prefix !== KEY_PREFIX) throw new Error(`unknown key format '${prefix}'`);
  if (!payload || !signature) throw new Error("empty payload or signature");
  return { claims: JSON.parse(fromBase64Url(payload).toString("utf8")), payload, signature };
};

/**
 * Is this key genuine, current, and for this version?
 *
 * Returns a reason instead of throwing, because every caller wants to say WHY a
 * key was refused: the CLI prints it, the app shows it, and a support reply is
 * mostly that one sentence. The Swift twin in apps/apple/Cupertino/License.swift
 * makes the same four checks in the same order.
 */
export const verifyKey = (key, { major, publicKey, revoked = [] }) => {
  let parsed;
  try {
    parsed = parse(key);
  } catch (error) {
    return { ok: false, reason: String(error?.message ?? error) };
  }

  const { claims, payload, signature } = parsed;
  const genuine = verify(
    null,
    Buffer.from(payload),
    publicKeyFrom(publicKey),
    fromBase64Url(signature),
  );
  if (!genuine) return { claims, ok: false, reason: "signature does not match" };
  if (revoked.includes(claims.id)) {
    return { claims, ok: false, reason: `licence ${claims.id} was revoked` };
  }
  if (major !== undefined && claims.major !== major) {
    return { claims, ok: false, reason: `key covers ${claims.major}.x, this build is ${major}.x` };
  }
  return { claims, ok: true, reason: "" };
};

File: apps/apple/Cupertino/License.swift
import CryptoKit
import Foundation

/// What a licence key says about itself, once it has proved it.
struct License {
  let id: String
  let email: String
  let major: Int
  let issuedAt: String
}

/// The answer, with the reason attached.
///
/// A bare `Bool` would be cheaper and would throw away the only part anyone
/// needs: *why*. The bridge relays this sentence to the MCP host, the menu bar
/// renders it, and a support reply is mostly this sentence — so it is produced
/// once, here, rather than reconstructed at each of the three.
enum LicenseCheck {
  case valid(License)
  case refused(String)

  var license: License? {
    if case .valid(let license) = self { return license }
    return nil
  }
}

/// Offline verification of a licence key.
///
/// The twin of `scripts/lib/license.mjs`, and deliberately a twin rather than a
/// shared library: the app cannot import JavaScript and the Worker cannot import
/// Swift. The two make the same four checks in the same order — format,
/// signature, revocation, version — and `scripts/lib/license.test.mjs` reads the
/// public key out of this file and asserts it matches, because the failure mode
/// of that one constant drifting is every paid key refused at once.
///
/// Nothing here reaches the network, and nothing here may start to.
/// `scripts/audit-network.sh` fails the build on `URLSession`, `getaddrinfo` and
/// the TLS entry points, and docs/licensing.md sells that on the front page.
/// CryptoKit trips none of them; an activation call would trip all of them.
///
/// Nor is any of this a tamper defence. `apps/apple/LICENSE` §1(c) grants
/// compiling, the source is public, and the gate this feeds is a dozen readable
/// lines in `ServerHost.swift` — so removing it is minutes of work for exactly
/// the audience this sells to, and no amount of hardening changes that. §2(b)
/// makes deleting the check a breach rather than a puzzle worth setting; the
/// deterrent is legal and social, never technical. This buys a key that cannot
/// be forged. It does not buy — and must not pretend to buy — a binary that
/// cannot be modified.
enum LicenseKey {
  /// Namespaces the format. A v2 key would carry a different one and be refused
  /// by name here rather than failing somewhere less legible.
  static let prefix = "cup1"

  /// The public half of the signing key: raw 32 bytes, base64url.
  ///
  /// Must equal `PUBLIC_KEY` in `scripts/lib/license.mjs`, which
  /// `scripts/lib/license.test.mjs` asserts by reading this file. Changing it
  /// here alone would refuse every key ever issued; changing it in both places
  /// would refuse every key issued before the change.
  static let publicKey = "_sGLrSm_Sg3bv2p0T8yfzelAvkEjAa2se9l2X4sgNA4"

  static func check(
    _ key: String?,
    major: Int = AppInfo.major,
    revoked: Set<String> = Revocations.ids
  ) -> LicenseCheck {
    guard let key, !key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      return .refused("no licence key")
    }

    let parts = key.trimmingCharacters(in: .whitespacesAndNewlines).split(
      separator: ".", omittingEmptySubsequences: false
    ).map(String.init)
    guard parts.count == 3 else { return .refused("expected three dot-separated parts") }
    guard parts[0] == prefix else { return .refused("unknown key format '\(parts[0])'") }
    guard !parts[1].isEmpty, !parts[2].isEmpty else {
      return .refused("empty payload or signature")
    }

    guard let payload = base64Url(parts[1]), let signature = base64Url(parts[2]) else {
      return .refused("payload or signature is not base64url")
    }
    guard let claims = try? JSONDecoder().decode(Claims.self, from: payload) else {
      return .refused("payload is not a licence")
    }

    guard let verifier = signer else {
      return .refused("this build has no signing key compiled in")
    }
    // Over the ENCODED payload, not the decoded claims: JSON key order and
    // whitespace then never have to agree across two languages.
    guard verifier.isValidSignature(signature, for: Data(parts[1].utf8)) else {
      return .refused("signature does not match")
    }

    // After the signature, never before. A forged id must not be waved through
    // by the simple expedient of not appearing on the list.
    guard !revoked.contains(claims.id) else {
      return .refused("licence \(claims.id) was revoked")
    }
    guard claims.major == major else {
      return .refused("key covers \(claims.major).x, this build is \(major).x")
    }

    return .valid(
      License(id: claims.id, email: claims.email, major: claims.major, issuedAt: claims.issuedAt))
  }

  // MARK: - Internals

  private struct Claims: Decodable {
    let id: String
    let email: String
    let major: Int
    let issuedAt: String
  }

  private static var signer: Curve25519.Signing.PublicKey? {
    guard let raw = base64Url(effectivePublicKey), !raw.isEmpty else { return nil }
    return try? Curve25519.Signing.PublicKey(rawRepresentation: raw)
  }

  private static var effectivePublicKey: String {
    #if DEBUG
      if let dev = developmentPublicKey() { return dev }
    #endif
    return publicKey
  }

  #if DEBUG
    /// The same `dev.json` `ServerLocator` already reads, with one more key:
    ///
    ///     { "node": …, "repo": …, "licensePublicKey": "voEH53_e3s…" }
    ///
    /// A development build has to be able to verify keys signed by a throwaway
    /// keypair, and the alternative — a `#if DEBUG` branch that skips the check
    /// — would mean the gate is never exercised by the builds it is developed
    /// against. This way Debug and Release run identical logic and differ only
    /// in which public key they trust. Compiled out of Release entirely, so it
    /// cannot become a way in.
    private static func developmentPublicKey() -> String? {
      let url = URL(fileURLWithPath: BridgeProtocol.socketDirectory)
        .appendingPathComponent("dev.json")
      guard
        let data = try? Data(contentsOf: url),
        let json = try? JSONSerialization.jsonObject(with: data) as? [String: String],
        let key = json["licensePublicKey"], !key.isEmpty
      else { return nil }
      return key
    }
  #endif

  private static func base64Url(_ text: String) -> Data? {
    var padded = text.replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    while padded.count % 4 != 0 { padded.append("=") }
    return Data(base64Encoded: padded)
  }
}
